// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: pb/hello/hello.proto

package hello

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on SayReq with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *SayReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SayReq with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in SayReqMultiError, or nil if none found.
func (m *SayReq) ValidateAll() error {
	return m.validate(true)
}

func (m *SayReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if l := utf8.RuneCountInString(m.GetMsg()); l < 1 || l > 10 {
		err := SayReqValidationError{
			field:  "Msg",
			reason: "value length must be between 1 and 10 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return SayReqMultiError(errors)
	}

	return nil
}

// SayReqMultiError is an error wrapping multiple validation errors returned by
// SayReq.ValidateAll() if the designated constraints aren't met.
type SayReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SayReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SayReqMultiError) AllErrors() []error { return m }

// SayReqValidationError is the validation error returned by SayReq.Validate if
// the designated constraints aren't met.
type SayReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SayReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SayReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SayReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SayReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SayReqValidationError) ErrorName() string { return "SayReqValidationError" }

// Error satisfies the builtin error interface
func (e SayReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSayReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SayReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SayReqValidationError{}

// Validate checks the field values on SayResp with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *SayResp) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SayResp with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in SayRespMultiError, or nil if none found.
func (m *SayResp) ValidateAll() error {
	return m.validate(true)
}

func (m *SayResp) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Msg

	if len(errors) > 0 {
		return SayRespMultiError(errors)
	}

	return nil
}

// SayRespMultiError is an error wrapping multiple validation errors returned
// by SayResp.ValidateAll() if the designated constraints aren't met.
type SayRespMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SayRespMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SayRespMultiError) AllErrors() []error { return m }

// SayRespValidationError is the validation error returned by SayResp.Validate
// if the designated constraints aren't met.
type SayRespValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SayRespValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SayRespValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SayRespValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SayRespValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SayRespValidationError) ErrorName() string { return "SayRespValidationError" }

// Error satisfies the builtin error interface
func (e SayRespValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSayResp.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SayRespValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SayRespValidationError{}
